File: client/src/types/analysis.ts
/** Stem information from analysis.json */
export interface StemInfo {
  name: string;
  paths: Record<string, string>; // speed (e.g., "1.0x") -> relative path
  hasNotes: boolean;
  peakDb: number;
}

/** Beat event from analysis.json */
export interface BeatEvent {
  time: number;
  type: "beat" | "downbeat";
  beatInMeasure: number | null;
}

/** Note information (for future features) */
export interface Note {
  start: number;
  end: number;
  pitch: number;
  velocity: number;
  pitchBend?: { time: number; cents: number }[];
}

/** Lyric word with timing */
export interface LyricWord {
  text: string;
  start: number;
  end: number;
  confidence: number;
}

/** Lyric line with words */
export interface LyricLine {
  text: string;
  start: number;
  end: number;
  words: LyricWord[];
}

/** Lyrics data */
export interface LyricsData {
  source: "lrc" | "txt" | "transcribed";
  lines: LyricLine[];
}

/** Full song analysis from analysis.json */
export interface SongAnalysis {
  title: string | null;
  artist: string | null;
  album: string | null;
  originalDuration: number;
  sampleRate: number;
  tempoBpm: number | null;
  timeSignature: [number, number] | null;
  stems: Record<string, StemInfo>;
  beats: BeatEvent[];
  notes?: Record<string, Note[]>;
  lyrics?: LyricsData | null;
  sourceFile: string;
  processingDate: string;
  converterVersion: string;
}

/** Summary for song browser */
export interface SongSummary {
  path: string;
  title: string | null;
  artist: string | null;
  duration: number;
  stemCount: number;
}

/** Available playback speeds */
export const PLAYBACK_SPEEDS = ["0.5x", "0.75x", "1.0x", "1.25x"] as const;
export type PlaybackSpeed = (typeof PLAYBACK_SPEEDS)[number];

/** Stem colors for UI */
export const STEM_COLORS: Record<string, string> = {
  drums: "bg-red-500",
  bass: "bg-blue-500",
  vocals: "bg-purple-500",
  guitar: "bg-green-500",
  piano: "bg-cyan-500",
  other: "bg-orange-500",
};

/** Get color for a stem, with fallback */
export function getStemColor(stemName: string): string {
  const lowerName = stemName.toLowerCase();
  for (const [key, color] of Object.entries(STEM_COLORS)) {
    if (lowerName.includes(key)) {
      return color;
    }
  }
  return STEM_COLORS.other;
}

File: src/music_tutor/models/analysis.py
"""Core analysis data models for Music Tutor.

These models represent the analysis output that gets serialized to JSON
alongside the separated stems.
"""

from dataclasses import dataclass, field
from typing import Literal


@dataclass
class PitchBendPoint:
    """A point in a pitch bend curve."""

    time: float  # seconds (relative to note start)
    cents: float  # deviation in cents from base pitch


@dataclass
class Note:
    """A detected note with timing and pitch information."""

    start: float  # seconds
    end: float  # seconds
    pitch: int  # MIDI note number (0-127)
    velocity: float  # 0.0-1.0 (from basic-pitch confidence)
    pitch_bend: list[PitchBendPoint] | None = None  # optional pitch bend curve


@dataclass
class BeatEvent:
    """A beat or downbeat event."""

    time: float  # seconds (at 1.0x speed)
    type: Literal["beat", "downbeat"]
    beat_in_measure: int | None = None  # 1, 2, 3, 4 for 4/4 time


@dataclass
class DrumStrike:
    """A detected drum hit from an isolated drum stem."""

    time: float  # seconds (at 1.0x speed)
    velocity: float  # 0.0-1.0 (relative amplitude)


@dataclass
class LyricWord:
    """A single word with timing."""

    text: str
    start: float  # seconds
    end: float  # seconds
    confidence: float  # 0.0-1.0 (from alignment model)


@dataclass
class LyricLine:
    """A line of lyrics with word-level timing."""

    text: str  # full line text
    start: float  # seconds
    end: float  # seconds
    words: list[LyricWord] = field(default_factory=list)


@dataclass
class LyricsData:
    """Word-level lyrics with timing."""

    source: Literal["lrc", "txt", "transcribed"]  # how lyrics were obtained
    lines: list[LyricLine] = field(default_factory=list)


@dataclass
class StemInfo:
    """Information about a separated stem."""

    name: str  # "vocals", "drums", "bass", "other"

    # Relative paths to audio files at each speed
    paths: dict[str, str] = field(default_factory=dict)  # {"1.0x": "stems/vocals_1.0x.flac", ...}

    # Whether pitch detection was run on this stem
    has_notes: bool = False

    # Peak loudness for UI normalization
    peak_db: float = 0.0


@dataclass
class SongAnalysis:
    """Root analysis object, serialized to JSON alongside stems."""

    # Metadata
    title: str | None
    artist: str | None
    album: str | None
    original_duration: float  # seconds
    sample_rate: int
    tempo_bpm: float | None = None
    time_signature: tuple[int, int] | None = None  # e.g., (4, 4)

    # Processing info
    source_file: str = ""  # original filename
    processing_date: str = ""  # ISO format
    converter_version: str = ""

    # Stem availability (relative paths from analysis file)
    stems: dict[str, StemInfo] = field(default_factory=dict)

    # Analysis data (at 1.0x speed - scale for other speeds)
    beats: list[BeatEvent] = field(default_factory=list)
    notes: dict[str, list[Note]] = field(default_factory=dict)  # keyed by stem name
    lyrics: LyricsData | None = None

    # Drum strike analysis (keyed by drum stem: kick, snare, hh, etc.)
    drum_strikes: dict[str, list[DrumStrike]] = field(default_factory=dict)
